cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +48,1538 @@ using parquet::detail::ColumnChunkDesc;
+  using hash_fn_type        = insert_hash_functor<T>;
+  using probing_scheme_type = cuco::linear_probing<SET_CG_SIZE, hash_fn_type>;
+
+  // Create a view of the hash set
+  auto hash_set_ref   = cuco::static_set_ref{cuco::empty_key{EMPTY_KEY_SENTINEL},
+                                           equality_fn_type{decoded_data},
+                                           probing_scheme_type{hash_fn_type{decoded_data}},
+                                           cuco::thread_scope_block,
+                                           storage_ref};
+  auto set_insert_ref = hash_set_ref.rebind_operators(cuco::insert);
+
+  // Decode values from the current dictionary page
+  for (auto value_idx = group.thread_rank(); value_idx < page.num_input_values;
+       value_idx += group.num_threads()) {
+    // Key (decoded value's global index) to insert into the cuco hash set
+    auto const insert_key = static_cast<key_type>(value_offset + value_idx);
+
+    // Decode the value from the page data
+    decoded_data[insert_key] =
+      decode_fixed_width_value<T>(page, chunk, value_idx, physical_type, error);
+
+    // Return early if an error has been set
+    if (is_error_set(error)) { return; }
+
+    // Insert the key (decoded value's global index) into the cuco hash set
+    set_insert_ref.insert(insert_key);
+  }
+}
+
+/**
+ * @brief Decode string column chunk dictionaries and evaluate (few) input predicates against
+ * decoded values
+ *
+ * @param pages Column chunk dictionary page headers
+ * @param results Span of device vector start pointers to store query results, one per predicate
+ * @param scalars Span of scalar device views, one per predicate
+ * @param operators Span of corresponding (in)equality operators, one per predicate
+ * @param total_num_scalars Total number of predicates
+ * @param total_row_groups Total number of row groups
+ * @param num_dictionary_columns Total number of columns with dictionaries
+ * @param dictionary_col_idx Index of the current dictionary column
+ * @param error Pointer to the kernel error code
+ */
+CUDF_KERNEL void __launch_bounds__(DECODE_BLOCK_SIZE)
+  evaluate_few_string_literals(PageInfo const* pages,
+                               cudf::device_span<bool*> results,
+                               ast::generic_scalar_device_view const* scalars,
+                               ast::ast_operator const* operators,
+                               cudf::size_type total_num_scalars,
+                               cudf::size_type total_row_groups,
+                               cudf::size_type num_dictionary_columns,
+                               cudf::size_type dictionary_col_idx,
+                               kernel_error::pointer error)
+{
+  // Each thread decodes values from one dictionary page and evaluates all input
+  // predicates against them
+
+  // Current row group index
+  for (size_type row_group_idx = cudf::detail::grid_1d::global_thread_id();
```

Comment:
Using one thread per dictionary page instead of (thread0 of) each warp per dictionary

Reply:
This doesn't seem to match up with the loop variable.  Is `row_group_idx` more like `row_group_chunk_idx` or `row_group_col_idx`?  That is, we're iterating over # row groups * # columns with dictionaries.

Reply:
Each thread will decode one dictionary page of the current column - one page per row group. The global `chunk_idx` we are currently operating on is computed at L828 taking care of `num_columns`